File: python-algorithm/leetcode/problem_138.py
"""138. Copy List with Random Pointer
https://leetcode.com/problems/copy-list-with-random-pointer/

A linked list is given such that each node contains an additional random
pointer which could point to any node in the list or null.

Return a deep copy of the list.

The Linked List is represented in the input/output as a list of n nodes.
Each node is represented as a pair of [val, random_index] where:

val: an integer representing Node.val
random_index: the index of the node (range from 0 to n-1) where random pointer
points to, or null if it does not point to any node.
"""


class Solution:

    def copy_random_list(self, head: 'Node') -> 'Node':
        if not head:
            return head
        store, index = {}, {}
        p = head
        i = 0
        while p:
            store[i] = Node(x=p.val)
            index[p] = i
            p = p.next
            i += 1
        p = head
        for j in range(i):
            if j < i - 1:
                store.get(j).next = store.get(j + 1)
            if p.random:
                store.get(j).random = store.get(index.get(p.random))
            p = p.next
        return store[0]

    def copy_random_list_2(self, head: 'Node') -> 'Node':
        """
        for-loop 2 times.
        time complexity: O(N)
        space complexity: O(N)
        """
        if not head:
            return head
        store = {}
        p = head
        while p:
            store[p] = Node(x=p.val)
            p = p.next
        p = head
        while p:
            if p.next:
                store.get(p).next = store.get(p.next)
            if p.random:
                store.get(p).random = store.get(p.random)
            p = p.next
        return store[head]

class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random

File: java-algo/src/main/java/com/leetcode/Problem138.java
package com.leetcode;

import java.util.HashMap;
import java.util.Map;

/**
 * 138. Copy List with Random Pointer
 * https://leetcode.com/problems/copy-list-with-random-pointer/
 * <p>
 * A linked list is given such that each node contains an additional random
 * pointer which could point to any node in the list or null.
 * <p>
 * Return a deep copy of the list.
 * <p>
 * The Linked List is represented in the input/output as a list of n nodes.
 * Each node is represented as a pair of [val, random_index] where:
 * <p>
 * val: an integer representing Node.val
 * random_index: the index of the node (range from 0 to n-1) where
 * random pointer points to, or null if it does not point to any node.
 */
public class Problem138 {
    public Node copyRandomList(Node head) {
        if (null == head)
            return null;
        Map<Integer, Node> store = new HashMap<>();
        Map<Node, Integer> index = new HashMap<>();
        Node p = head;
        int i = 0;
        while (p != null) {
            store.put(i, new Node(p.val));
            index.put(p, i);
            p = p.next;
            i++;
        }
        p = head;
        for (int j = 0; j < i; j++) {
            if (j < i - 1) {
                store.get(j).next = store.get(j + 1);
            }
            if (p.random != null) {
                store.get(j).random = store.get(index.get(p.random));
            }
            p = p.next;
        }

        return store.get(0);
    }

    public Node copyRandomList2(Node head) {
        if (null == head)
            return null;
        Map<Node, Node> store = new HashMap<>();
        Node p = head;
        while (p != null) {
            store.put(p, new Node(p.val));
            p = p.next;
        }
        p = head;
        while (p != null) {
            if (p.next != null)
                store.get(p).next = store.get(p.next);
            if (p.random != null)
                store.get(p).random = store.get(p.random);
            p = p.next;
        }
        return store.get(head);
    }

    static class Node {
        int val;
        Node next;
        Node random;

        public Node(int val) {
            this.val = val;
            this.next = null;
            this.random = null;
        }
    }

}
